feat: dropdown

svifty7 opened the pull request.

Kossesor committed:
refactor(AbilityTable): typos

Kossesor committed:
refactor(AbilityRaces): script setup

Kossesor committed:
refactor(AbilityCalcView): script setup

Kossesor committed:
feat(ui-dropdown): dropdown input

Kossesor committed:
Merge branch 'dev' into feature/dropdown

# Conflicts:
#	src/views/Tools/AbilityCalc/AbilityCalcView.vue

Kossesor committed:
feat(dropdown): focused toggle via icon

Kossesor committed:
feat(dropdown): togglers and placeholder

Kossesor committed:
Merge remote-tracking branch 'origin/dev' into feature/dropdown

Kossesor committed:
feat(dropdown): styling wrapper

Kossesor committed:
feat(dropdown): styles for active dropdown

Kossesor committed:
fix: idea import optimization

Kossesor committed:
feat: completed toggle for select

Kossesor committed:
feat: filter options for select

Kossesor committed:
fix: style fixes

Kossesor committed:
fix: watch focus to clear filter

Kossesor committed:
feat: v-model for dropdown

Kossesor committed:
Merge branch 'feature/uikit-search' into feature/dropdown

# Conflicts:
#	src/common/helpers/download.ts
#	src/components/UI/menu/NavSearch.vue
#	src/features/bookmarks/components/CustomBookmarks/CustomBookmarkCategory.vue
#	src/features/bookmarks/components/CustomBookmarks/CustomBookmarkGroup.vue
#	src/features/filter/ListFilter.vue
#	src/pages/Tools/AbilityCalc/AbilityCalcView.vue
#	src/pages/Tools/AbilityCalc/AbilityRaces/AbilityRaces.vue
#	src/pages/Tools/AbilityCalc/AbilityTable.vue
#	src/shared/ui/kit/UiDropdown.vue
#	src/shared/ui/modals/AuthModal.vue
#	src/shared/ui/modals/BaseModal.vue
#	src/types/Workshop/Bestiary.types.ts
#	src/views/IndexView.vue

Kossesor committed:
Feat: selected options rendering

Kossesor committed:
Feat: added disabled prop and improved classes

Kossesor committed:
fix: bugfix with filter hasnt been cleared

Kossesor committed:
fix: removing unnecessary onSearch func

Kossesor committed:
fix: input full height

olegvolkov91 commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,366 @@
+  );
+
+  const toggleIcon = computed(() =>
+    focused.value ? 'arrow/up' : 'arrow/down'
```
чисто вкусовщина => `arrow/${focused.value ? 'up' : 'down'}`

Kossesor committed:
fix: full width and text wrapping

UrSenpaiii commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,366 @@
+  import UiInput from '@/shared/ui/kit/UiInput.vue';
+
+  const props = defineProps({
+    modelValue: {
```
крутяк, не знал об этой штуке. Переделал на нее

UrSenpaiii commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,366 @@
+
+<script setup lang="ts">
+  import { onClickOutside } from '@vueuse/core';
+  import { xor } from 'lodash';
```
это idea всё сама там правила)

Kossesor committed:
fix: using defineModel now and resetting modelValue now

Kossesor committed:
refactor: added ts types

Kossesor committed:
refactor: Option type

Kossesor committed:
feat: isSearchable prop and slotted layout

Kossesor committed:
feat: any object as an option via label and trackBy props

Kossesor committed:
refactor: AbilityRaces.vue script to setup and implementation of dropdown

Kossesor committed:
refactor: rename dropdown to multiselect

Kossesor committed:
feat: added left/right slots for select

Kossesor committed:
refactor: implementation of new select

svifty7 commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,426 @@
+  const modelValue = defineModel();
+
+  const props = defineProps({
+    trackBy: {
+      type: Function,
+      required: true
+    },
+    label: {
+      type: Function,
+      required: true
+    },
+    options: {
+      type: Array,
+      default: () => []
+    },
+    placeholder: {
+      type: String,
+      default: ''
+    },
+    disabled: {
+      type: Boolean,
+      default: false
+    },
+    isMultiple: {
+      type: Boolean,
+      default: false
+    },
+    isSearchable: {
+      type: Boolean,
+      default: false
+    }
```
лучше переделать на 
```typescript
defineProps<{
  label: string;
  trackBy: string;
  options: Array<T>;
  disabled?: boolean;
  isMultiple?: boolean;
  isSearchable?: boolean;
}>()
```

UrSenpaiii replied:
А шо, так можно было, ДА?)

svifty7 replied:
конечно)) мы все так новое делаем))

svifty7 commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,426 @@
+  });
+
+  const input = ref<typeof UiInput | null>(null);
+  const dropdownHeader = ref<HTMLDivElement | null>(null);
```
вкусовщина, но так кода меньше)
```ts
const dropdownHeader = ref<HTMLDivElement>();
```

svifty7 commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,426 @@
+
+  const input = ref<typeof UiInput | null>(null);
+  const dropdownHeader = ref<HTMLDivElement | null>(null);
+  const focused = ref<Boolean>(false);
+  const filter = ref<string>('');
```
тут типизация уже излишняя

UrSenpaiii replied:
По поводу типизайций там еще нужно будет поглядеть

svifty7 replied:
если ты указываешь строку как значение по умолчанию, то он автоматически присваивает тип)

svifty7 commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,426 @@
+  const dropdownHeader = ref<HTMLDivElement | null>(null);
+  const focused = ref<Boolean>(false);
+  const filter = ref<string>('');
+  const selectedOptions = ref<Array<Object>>([]);
```
вместо `Object` нужно описать тип или использовать generic `T`

svifty7 commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,426 @@
+  </div>
+</template>
+
+<script setup lang="ts">
```
можно добавить generic `generic="T extends Record<string, unknown>"`

svifty7 commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,426 @@
+  import SvgIcon from '@/shared/ui/icons/SvgIcon.vue';
+  import UiInput from '@/shared/ui/kit/UiInput.vue';
+
+  const modelValue = defineModel();
```
типизируем из дженерика `T`

svifty7 commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,426 @@
+    focused.value ? 'arrow/up' : 'arrow/down'
+  );
+
+  const isSelectedClass = (option: Object) => {
+    return selectedOptions.value.filter(
+      o => props.trackBy(o) === props.trackBy(option)
+    ).length;
+  };
```
1. аналогично типизируем `Object`, в остальных местах тоже учесть нужно
2. `trackBy` лучше сделать строкой, и использовать как `option[props.trackBy]` (с `props.label` такая же фигня - учесть везде надо))

UrSenpaiii replied:
Вот тут проблемка, оно не просто так функция.
Что делать, если trackBy или option нужны свойства второго и более уровня?
К примеру ```race.name.rus```

svifty7 replied:
первый способ для типизации лучше

svifty7 commented on src/shared/ui/kit/UiDropdown.vue:
```diff
@@ -0,0 +1,426 @@
+    return optionsToDisplay.map(el => props.label(el)).join(', ');
+  });
+
+  watch(focused, f => {
+    if (!f) {
+      filter.value = '';
+    }
+  });
+
+  watch(modelValue, mv => {
+    if (!mv) {
+      selectedOptions.value = [];
+    }
+  });
```
однобуквенные переменные потом усложнят сильно понимание кода, нужно переписать на осмысленные, либо `value`))

Kossesor committed:
refactor(UiDropdown): props define

Kossesor committed:
refactor(UiDropdown): add generic and renaming for some funcs

Kossesor committed:
refactor(UiDropdown): trackBy and label change from funcs to string type via lodash get

Kossesor committed:
fix(uiMultiselect): remove unnecessary options in slots

Kossesor committed:
refactor: default choiceDouble; direct modelValue mutation

resetting values on choiceDouble
add default choiceDouble
remove selectedOptions from uiMultiselect

Kossesor committed:
merge: dev into feature/dropdown

Kossesor committed:
refactor(AbilityPointBuy): Composition API and UiMultiselect integration

Kossesor committed:
refactor(AbilityCalcView): Composition API

Kossesor committed:
refactor(AbilityRandom): Composition API

Kossesor committed:
refactor(AbilityRandom): UiMultiselect integration

Kossesor committed:
fix(AbilityRandom): Multiselect placeholder improvement and AbilityRandom fix

Kossesor committed:
refactor(AbilityArray): Composition API

Kossesor committed:
refactor(AbilityArray): UiMultiselect integration. Bug fixes in prev components

Kossesor committed:
fix(AbilityRandom): reselect already selected option

Kossesor committed:
fix(TreasuryView): httpClient

Kossesor committed:
refactor(TreasuryView): UiMultiselect integration and types fixes

Kossesor committed:
refactor(TraderView): UiMultiselect integration

Kossesor committed:
fix(UiMultiselect): component incorrect display in some cases; improved BEM

Kossesor committed:
fix(UiMultiselect): comparison fixes in isSelected class and filteredOptions

Kossesor committed:
chore(Ability): remove unused classes

Kossesor committed:
refactor(YoutubeView): UiMultiselect integration

Kossesor committed:
fix(UiMultiselect): isSelectedClass finishing

Kossesor committed:
refactor(EncountersView): Composition API and UiMultiselect integration

Kossesor committed:
fix(UiMultiselect): autofocus on input

Kossesor committed:
refactor(UiMultiselect): header label from slot to prop

Kossesor committed:
feat(UiMultiselect): add render of grouped options

Kossesor committed:
refactor(ClassDetail): UiMultiselect integration

Kossesor committed:
refactor(UiMultiselect): class

svifty7 commented on src/pages/tools/ability-calc/ability-races/AbilityRaces.vue:
```diff
@@ -276,75 +223,47 @@
 
-      <ui-select
+      <ui-multiselect
+        v-model="selectedChoiceDouble"
```
по всему МРу пересмотри, пожалуйста, зачем `v-model`, когда есть `@update:model-value`, и наоборот?
может стоит от чего-то отказаться? сразу скажу, что `v-model` через `computed(() => {get: () => {}, set: (v) => {}})` так себе затея, читать потом сложнее))

UrSenpaiii replied:
Пока чуть выше тебе отписывал, как раз словил себя на этой же мысле 
> Т.е. там два варианта, мы передаем model-value и он просто обновляется сам или же ловим ивент update:model-value и делаем что-то более специфическое с выбранными данными
Следовательно сейчас упрощу, где можно

svifty7 commented on src/pages/tools/encounters/EncountersView.vue:
```diff
@@ -1,168 +1,127 @@
+  onBeforeMount(() => getOptions());
+
+  // eslint-disable-next-line func-names
+  const sendForm = throttle(async function () {
```
тут можно сделать стрелочную функцию вместо `async function ()` и после этого убрать коммент для отключения линтера)

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+    trackBy: string;
+    options: Array<T>;
+    disabled?: boolean;
+    headerLabel?: string | number;
```
Мне кажется, стоит убрать этот пропс. Вместо него надо создать компонент `<ui-form-item/>` с пропсом И слотом `label`

UrSenpaiii replied:
Вообще изначально реализовал как и в прошлом селекте через `slot`, а вот `placeholder` сразу через `props`.
И вот пока писал, думал, делать ли `placeholder` тоже через `slot`, как было в прошлом или переводить `label` в `prop`, а может и то и другое сделать. Но так как за всё время `slot` был излишен - сделал через `prop`.

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+  const modelValue = defineModel<T>();
+
+  const props = defineProps<{
+    label: string;
+    trackBy: string;
```
а эти два пропса переименовать в `label-key`, `value-key`

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+    disabled?: boolean;
+    headerLabel?: string | number;
+    placeholder: string | number;
+    isMultiple?: boolean;
+    isSearchable?: boolean;
+    isGrouped?: boolean;
```
пропсы без `is` лучше будут смотреться, наверное, а последний можно переименовать в `grouping`

UrSenpaiii replied:
Вообще это правило именования булевых переменных. А еще раньше чтобы вместо `<div is-searchable="true" />` можно было указать просто `<div is-searchable />` обязательно нужен был префикс is в названии.

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+    isMultiple?: boolean;
+    isSearchable?: boolean;
+    isGrouped?: boolean;
+    groupLabel?: string;
+    groupValues?: string;
```
эти пропсы для лучшего понимания стоит закончить на `Key`

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+    groupValues?: string;
+  }>();
+
+  const input = ref<typeof UiInput | null>(null);
```
можно опустить оба `null` и проверять переменную на `undefined`

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+
+  const input = ref<typeof UiInput | null>(null);
+  const header = ref<HTMLDivElement>();
+  const focused = ref<Boolean>(false);
```
тут можно или опустить излишнюю типизацацию, или сделать её примитивом `boolean` вместо конструктора)

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+  const focused = ref<Boolean>(false);
+  const filter = ref<string>('');
+
+  onClickOutside(header, () => {
+    focused.value = false;
+  });
+
+  const toggleFocus = (e: Event) => {
+    if (e.type === 'focusin') focused.value = true;
+
+    if (e.type === 'click') {
+      focused.value = !focused.value;
+
+      if (focused.value) input.value?.focusInput();
+    }
```
слей себе последний дев и пройдись линтером)))

svifty7 replied:
посмотри у vueuse композиции для отслеживания состояния фокуса, возможно это поможет сократить код и не придется поддерживать самим состояние фокуса?

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+    focused.value ? 'arrow/up' : 'arrow/down',
+  );
+
+  const isSelectedClass = (option: T) =>
```
почему `Class`?

UrSenpaiii replied:
Рудимент, просто изначально юзается для навешивания css класса

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+
+  const isSelectedClass = (option: T) =>
+    String(get(option, props.label)) === String(props.placeholder) ||
+    isEqual(modelValue.value, option);
```
может стоит проверять по `option[props.valueKey]`?

UrSenpaiii replied:
значения `option` нужно брать только через `get()`, так как `option[key]` не работает в случае, если свойтв больше, чем одно. `get()` делает deep search так сказать)

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+  };
+
+  const displaySelectedOptions = computed(() =>
+    get(modelValue.value, props.label),
```
если используешь данный хелпер, то думаю стоит по всему компоненту в похожих местах им пользоваться, а не только в 1-2 случаях)

UrSenpaiii replied:
Этот именно для реактивного рендера `SelectedOptions`. а так, `get()` используется немного с разными аргументами, да и сам по себе он вполне лаконичен

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+    if (!isFocused) filter.value = '';
+  });
+
+  const filteredOptions = computed(() => {
+    let newOptions: Array<T> = [];
+
+    if (props.isGrouped) {
+      newOptions = props.options.reduce(
+        (res, el) => [
+          ...res,
+          { ...el[props.groupLabel], groupLabel: true },
+          ...el[props.groupValues],
+        ],
+        newOptions,
+      );
+    } else {
+      newOptions = props.options.filter((el) => {
+        const optionLabel = String(get(el, props.label)).toUpperCase();
+        const currentFilter = String(filter.value).toUpperCase();
+
+        return optionLabel.includes(currentFilter);
+      });
+    }
+
+    return newOptions;
+  });
```
предлагаю вот такой вариант вместо переменных и `else`)
```typescript
const filteredOptions = computed<Array<T>>(() => {
  if (props.isGrouped) {
    return props.options.reduce(
      (res, el) => [
        ...res,
        { ...el[props.groupLabel], groupLabel: true },
        ...el[props.groupValues],
      ],
     [],
    );
  }
     
  return props.options.filter((el) => {
    const optionLabel = String(get(el, props.label)).toUpperCase();
    const currentFilter = String(filter.value).toUpperCase();

    return optionLabel.includes(currentFilter);
  });
});
```

UrSenpaiii replied:
Буквально у друзей в дс консультировался как лучше)

вот такой вот вариант был)
`
  const filteredOptions = computed(() =>
    props.isGrouped
      ? props.options.reduce(
          (res, el) => [
            ...res,
            { ...el[props.groupLabelKey], groupLabel: true },
            ...el[props.groupValuesKey],
          ],
          newOptions,
        )
      : props.options.filter((el) => {
          const optionLabel = String(get(el, props.labelKey)).toUpperCase();
          const currentFilter = String(filter.value).toUpperCase();

          return optionLabel.includes(currentFilter);
        }),
  );
`

А так я просто придерживаюсь одного `return`, мол одна точка выхода их кода, блаблабла)

svifty7 commented on src/shared/ui/kit/UiMultiselect.vue:
```diff
@@ -0,0 +1,393 @@
+    isEqual(modelValue.value, option);
+
+  const selectOption = (option: T) => {
+    modelValue.value = option;
```
тут стоит использовать `cloneDeep`

UrSenpaiii commented on src/pages/admin/youtube/YoutubeView.vue:
```diff
@@ -84,11 +84,11 @@
             <span>Количество видео на странице</span>
 
-            <ui-select
-              :model-value="itemsPerPage.find((item) => item.value === size)"
```
При селекте опшина из списка он так или иначе триггерит `update:model-value` и передаёт туда выбранное значение, которое мы накидываем на `size`.
Ну как бы можно сделать прослойку в виде переменной по типу `selectedPageSize`, но зачем?
Т.е. там два варианта, мы передаем `model-value` и он просто обновляется сам или же ловим ивент `update:model-value` и делаем что-то более специфическое с выбранными данными

UrSenpaiii commented on src/pages/character/classes/classes-detail/ClassDetail.vue:
```diff
@@ -353,29 +353,17 @@
-            :model-value="currentSelectArchetype"
-            :options="currentArchetypes"
+          <ui-multiselect
+            v-model="currentSelectArchetype"
```
А поточнее?

UrSenpaiii commented on src/pages/character/classes/classes-detail/ClassDetail.vue:
```diff
@@ -104,7 +105,7 @@
   });
 
-  const currentSelectArchetype = computed(() => {
+  const getSelectedArchetype = () => {
```
да он какой-то нагроможденный, блин(

UrSenpaiii commented on src/pages/tools/ability-calc/AbilityRandom.vue:
```diff
@@ -156,36 +142,28 @@
 
-        <template #singleLabel>
-          {{ roll.name || 'Выбрать хар-ку' }}
+        <template #option="{ name, key }">
```
Да я вообще думал, может реализовать вот эту синию точку внутри компонента, т.к. при мультивыборе можно как раз ей отмечать выбранный пул элементов. Как думаешь?

P.S дада, мультиселект не реализован, т.к еще нигде не использовался на сайте)

UrSenpaiii commented on src/pages/tools/ability-calc/ability-races/AbilityRaces.vue:
```diff
@@ -276,75 +223,47 @@
 
     <div class="ability-races__fields">
-      <ui-select
-        :custom-label="({ name: { rus } }) => rus"
```
что именно? по функции?
Я как раз так изначально и сделал, мы по итогу решили сменить на текущий вариант. Ну и я не совсем понимаю, зачем

Kossesor committed:
fix: types fixes

Kossesor committed:
refactor(UiMultiselect): props rename
